File: trees/2nd_largest_element_BST.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

// tree node
class Node {
    public:
        int data;
        Node* left;
        Node* right;

        // constructor
        Node(int data) {
            this->data = data;
            this->left = NULL;
            this->right = NULL;
        }
};

class Tree {

    public:
        Node* root;

        // constructor
        Tree() {
            this->root = NULL;
        }

        void construct_tree() {

            //        50
            //        / \
            //       /   \
            //      /     \
            //     /       \
            //    10       60
            //   /  \     /  \
            //  5   20  55   70
            //          /   /  \
            //        51   65  80
            Node* root = new Node(50);
            root->left = new Node (40);
            root->left->right = new Node (45);
            root->left->left = new Node (30);
            root->left->left->left = new Node (20);
            //Node* temp = root;

            //temp->left = new Node(10);
            //temp->right = new Node(60);

            //temp = temp->left;
            //temp->left = new Node(5);
            //temp->right = new Node(20);

            //temp = root->right;
            //temp->left = new Node(55);
            //temp->right = new Node(70);

            //temp->left->left = new Node(51);
            //temp->right->left = new Node(65);
            //temp->right->right = new Node(80);

            this->root = root;
            return;
        }

        int get2ndLargestInBSTInorder() {
            int node_count = 0;
            return this->get2ndLargestInBSTInorderUtil(this->root, node_count);
        }

        void get2ndLargestInBST() {
            // a temp variable, so that the original root value does not change
            Node* root = this->root;

            // root cannot be null and must have at least 1 child
            if((root == NULL) || (root->right == NULL && root->left == NULL)) {
                cout << "Cannot find the 2nd largest element" << endl;
                return;
            }

            // if tree does not have a right subtree, then largest in left subtree
            if(root->left && (root->right == NULL)) {
                Node* secondLargest = this->getRightMostNode(root->left);
                cout << "2nd largest element = " << secondLargest->data << endl;
                return;
            }

            while(root) {
                // if current is the parent of the largest element, and the largest has no children
                if(root->right && (root->right->left == NULL) && (root->right->right == NULL)) {
                    cout << "2nd largest element = " << root->data << endl;
                    return;
                }

                root = root->right;
            }

            cout << "1 Cannot find the 2nd largest element" << endl;
            return;
        }

    private:
        // helper function for get2ndLargestInBSTInorder
        int get2ndLargestInBSTInorderUtil(Node* root, int node_count) {
            if(root == NULL) {
                return node_count;
            }

            node_count = get2ndLargestInBSTInorderUtil(root->right, node_count);

            node_count++;
            if(node_count == 2) {
                cout << "2nd largest element = " << root->data << endl;
            }

            node_count = get2ndLargestInBSTInorderUtil(root->left, node_count);
            return node_count;
        }

        Node* getRightMostNode(Node* root) {
            if(root == NULL) {
                return NULL;
            }

            if(root->right == NULL) {
                return root;
            }
            return getRightMostNode(root->right);
        }
};

int main() {

    Tree tree;
    tree.construct_tree();

    // Method1: Get inorder, and 2nd node in reverse inorder
    // Running time O(n), Space O(1)
    tree.get2ndLargestInBSTInorder();

    // Method2: Find the largest node, and then the second largest
    // Running time O(n), Space O(1)
    tree.get2ndLargestInBST();

    return 0;
}

File: trees/trees_py/2nd_largest_element_BST.py
"""
Find the 2nd largest element in a BST

The largest element lies on the right most node
The 2nd largest element lies on the left of the right most element

You can do an inOrder traversal and keep the last largest element
"""

# -*- encoding: utf-8 *-*


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class Tree:
    def __init__(self):
        self.root = None

    #           7
    #          / \
    #         /   \
    #        /     \
    #       /       \
    #      /         \
    #     3          11
    #   /  \         / \
    #  /    \       /   \
    # 1      5     9    13
    #  \    / \   / \   / \
    #   2  4   6 8  10 12 14

    def construct_tree(self):
        self.root = Node(7)
        self.root.left = Node(3)
        self.root.left.left = Node(1)
        self.root.left.left.right = Node(2)
        self.root.left.right = Node(5)
        self.root.left.right.left = Node(4)
        self.root.left.right.right = Node(6)
        self.root.right = Node(11)
        self.root.right.left = Node(9)
        self.root.right.left.left = Node(8)
        self.root.right.left.right = Node(10)
        self.root.right.right = Node(13)
        self.root.right.right.left = Node(12)
        self.root.right.right.right = Node(14)

    def in_order_traversal(self, root):
        if root is None:
            return
        self.in_order_traversal(root.left)
        print(root.data, end=" ")
        self.in_order_traversal(root.right)

    def get_node_count(self, root):
        if root is None:
            return 0
        return 1 + self.get_node_count(root.left) + self.get_node_count(root.right)

    def right_most_node(self, root):
        if root is None:
            return
        while(root.right):
            root = root.right
        return root

    def second_largest_element(self, root):
        if (root is None) or (root.left is None and root.right is None):
            return None

        curr_node = root
        while curr_node:
            # If tree does not have a right subtree, find element in left subtree
            if curr_node.left and root.right is None:
                return self.right_most_node(curr_node.left)

            if curr_node.right and curr_node.right.left is None and curr_node.right.right is None:
                return curr_node

            curr_node = curr_node.right

        return


def main():
    tree = Tree()
    tree.construct_tree()

    print("in order traversal: ", end=" ")
    tree.in_order_traversal(tree.root)
    print()

    print("total nodes in tree - ", tree.get_node_count(tree.root))
    print()

    print("right most node in tree", end=" ")
    right_most_node = tree.right_most_node(tree.root)
    if right_most_node:
        print(right_most_node.data)
    else:
        print("Cannot find right most node in tree")
    print()

    print("2nd largest element in tree", end=" ")
    second_largest_element = tree.second_largest_element(tree.root)
    if second_largest_element:
        print(second_largest_element.data)
    print()
    return

if __name__ == '__main__':
    main()
